pint/formatting.py

```diff
@@ -287,10 +291,64 @@ def format_compact(unit: UnitsContainer, registry: UnitRegistry, **options) -> s
         division_fmt="/",
         power_fmt="{}**{}",
         parentheses_fmt=r"({})",
+        registry=registry,
         **options,
     )
 
 
+dim_order = [
+    "[substance]",
+    "[mass]",
+    "[current]",
+    "[luminosity]",
+    "[length]",
+    "[time]",
+    "[temperature]",
+]
+
+
+def dim_sort(units: Iterable[list[str]], registry: UnitRegistry):
+    """Sort a list of units by dimensional order.
+
+    Parameters
+    ----------
+    units : list
+        a list of unit names (without values).
+    registry : UnitRegistry
+        the registry to use for looking up the dimensions of each unit.
+
+    Returns
+    -------
+    list
+        the list of units sorted by most significant dimension first.
+
+    Raises
+    ------
+    KeyError
+        If unit cannot be found in the registry.
+    """
+    ret_dict = dict()
+    len(units) > 1
+    for name in units:
+        cname = registry.get_name(name)
+        if not cname:
+            continue
+        dim_types = iter(dim_order)
+        while True:
+            try:
+                dim = next(dim_types)
+                if dim in registry.get_dimensionality(cname):
+                    if dim not in ret_dict:
+                        ret_dict[dim] = list()
+                    ret_dict[dim].append(cname)
+                    break
+            except StopIteration:
+                raise KeyError(f"Unit {cname} has no recognized dimensions")
+
+    ret = sum([ret_dict[dim] for dim in dim_order if dim in ret_dict], [])
```

Comment:
I wonder if you could use the `sorted` builtin for something like this? Like, using the `key` parameter to translate a unit to an integer position in `dim_order`, with unknown dimensions being always last (or first)? The advantage would be that it would be easier to understand, and it might also be a bit faster?

Reply:
I had thought about it, but other problems with the implementation led me in a different direction (which now passes the doc tests and reveals a potential problem in the tutorial).  The long and short is that `cname` is great for finding dimensionality, but test cases rely on leaving names as they are (such as `m` not being renamed `meter`).  Also I had hoped to use the natural sorting of positive and negative exponents as a way to process shorter lists of units (possibly skipping any sorting if we just had singleton terms, such as `m` or `s ** -2`).  This proved problematic, and so the implementation now minimally disturbs the names and tuples of the incoming units.  I think that the extra work of setting up a sortable key will overcome whatever other simplifications that `sorted` might provide.